man/man1/mergeclm.1
.TH mergeclm 1  "USER COMMANDS"
.SH NAME
mergeclm \- merge LPJmL climate data files
.SH SYNOPSIS
.B mergeclm
[\-v] [\-f] [\-longheader] [\-size4]
\fIinfile\fP ... \fIoutfile\fP
.SH DESCRIPTION
Program merges bands of climate data files in clm format into one clm file.
.SH OPTIONS
.TP
\-v,\--verbose
Print filename and number of bands of each input file.
.TP
\-f
Force overwriting of existing output file.
.TP
\-longheader
Long (version 2) header assumed for clm files.
.TP
\-size4
Size of data is set to 4 bytes for version 2 clm files, default is 2 bytes.
.TP
.I infile
Filename(s) of clm files to paste. If filename is "zero" one band containing all zero values is added.
.TP
.I outfile
Filename of merged clm file.
.SH EXAMPLE
.TP
Merge clm data file from 12 clm files:
.B mergeclm
temp_1901-2003_01.clm temp_1901-2003_02.clm temp_1901-2003_03.clm temp_1901-2003_04.clm temp_1901-2003_05.clm temp_1901-2003_06.clm temp_1901-2003_07.clm temp_1901-2003_08.clm temp_1901-2003_09.clm temp_1901-2003_10.clm temp_1901-2003_11.clm temp_1901-2003_12.clm temp_1901-2003.clm
.PP
.SH EXIT STATUS
.B mergeclm
returns a zero exit status if the clm file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
lpjml(1), cutclm(1), splitclm(1), txt2clm(1), grid2clm(1), cft2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)
- code review:

### Added

- Check for valid `swc_bnf` parameters added in `fscanpftpar.c`.
- Wirth et al. (2024)  article added to `REFERENCES`.

### Changed

- `phi_bnf` parameter calculated from `swc_bnf` parameter instead of reading from `pft.cjson` file in `fscanpftpar.c`.

### Removed

- Parameter `phi_bnf` removed from `pft.cjson`.


## [5.10.0] - 2025-08-14

### Contributors

- author: Jens Heinke (heinke@pik-potsdam.de), Sibyll Schaphoff (sibylls@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- option to specify year after which deforestation is no longer feeding product pools but biomass is burnt instead: `luc_burn_startyear` (disabled)
- option to specify costs for N recovery upon turnover (pre-compiler switch `NRECOVERY_COST`, disabled)

### Changed

- parameter changes: 
  - `alphaa` from 0.5 to 0.7 for natural vegetation; 
  - `theta` from 0.9 to 0.95; 
  - `fn_turnover` PFT1: 0.5, PFT2: 0.5, PFT3: 0.7, PFT4: 0.3, PFT5: 0.3, PFT6: 0.7, PFT7: 0.3, PFT8: 0.3; updated values from [Sophia et al. 2024](https://doi.org/10.5194/bg-21-4169-2024) for all other PFTs except crops
  - `cnratio_leaf.low/median/high` for all PFTs updated from TRY data
  - `b` (ratio of dark respiration Rd to maximum photosynthetic capacity Vcmax) set to 0.031 for C3 and C4 plants (value for 25°C, from [Wang et al. 2020](https://doi.org/10.1111/gcb.14980), consistent with TRY data)
  - `aphen_min` from 60 to 90 and `aphen_max` from 245 to 180
  - `nfixpot` from 0.01 to 0.5 (middle of proposed range in [Yu and Zhuang 2020](https://doi.org/10.5194/bg-17-3643-2020))
  - `turnover` parameter for leaves&roots set to 2 for TeNE and 3 BoNE 
  - temperature boundary between temperate and boreal zone changed from -2°C to -5°C
- acclimation of `b` to mean vegetation period temperature (following [Wang et al. 2020](https://doi.org/10.1111/gcb.14980))
- calculation of structural leaf N content (leaf N independent of Vcmax) based on `ncleaf.low` in both `ndemand_xx` and `vmaxlimit_xx` (was `ncleaf.medium` in `ndemand_xx`)
- scaling of vmax with f_LAI removed
- N recovery from turnover directly added to `bm_inc.nitrogen`
- N uptake now separated for uptake of NH4 and NO3 with parameters from  [Craig et al. 2025](https://doi.org/10.1111/nph.70140) (median of values)
- trees are burnt (instead of added to the litter) when running with `luc_timber=FALSE`
- beginning of vegetative period for tropical raingreen trees set at beginning of 6 month period with highest precipitation sum
- leaf shedding for deciduous trees induced after `aphen_max` or when `phen<0.1` after `aphen_min`
- `phen` permanently set to 1 for tropical evergreen trees

### Removed

### Fixed

- corrected conversion of vmax in computation of N demand, which was incorrectly scaled with `24/daylength`
- corrected temperature sensitivity of N demand (`k_temp` changed from 0.02 to 0.0693)
- corrected `f_NCplant` to comply with equation from [Smith et al. 2014](https://doi.org/10.5194/bg-11-2027-2014)
- corrected phenology of trees:
  - leaf turnover of raingreen trees at leaf shedding (same as for summergreen trees, had continuous daily leaf turnover like evergreen trees)
  - `phen` set to zero at leaf shedding until restart of phenological cycle to prevent phen>1 without leaf carbon 


## [5.9.28] - 2025-07-31

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Maik Billing (billing@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de), Sebastian Ostberg(ostberg@pik-potsdam.de), Jens Heinke (heinke@pik-potsdam.de)

### Added

- Option `-h` added to utilities `addheader`, `cdf2bin`, `regridclm`, `regridsoil` and `printclm` to print help text.
- Notice added in `configure.sh` that `make clean; make all` has to be performed after change in configuration.
- Support for short datatype added in `cmpbin` utility.
- Shell script `allbin2cdf` added to convert all binary raw output files in a directory to NetCDF files.
- Option `-nounit` added to `bin2cdf` to set unit if unit is defined in the metafile as an empty string.

### Changed

- If option `-ncell 0` is used in `addheader` utility then the number of cells is calculated from the file size of the binary file.
- Utilities `grid2clm` and `cft2clm` are replaced by alias to `addheader`.
- File `default.md` converted from DOS to Linux format.
- If scaling factor is set as an option in `bin2cdf` then this scaling factor is used insteads that one defined in the JSON metafile.
- Man pages updated.

### Fixed

- Missing `free(axis->comment)` added in `freeaxis()` to avoid memory leak.
- Missing deallocation of list added in `open_config.c`.
- Function `nc_close()` replaced by `closeclimate_netcdf()` in `cdf2clm.c` and `cdf2bin.c` to avoid memory leak.
- Argument for error message corrected in `getcellindex.c` and `getcountry.c`.
- Typo in man page of `configure.sh` fixed.
- Comment for `pft` initialized in `initsetting_netcdf.c`.
- Access to undefined `pft_name.comment` removed in `bin2cdf.c` to avoid SEGV.

src/utils/mergeclm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                         m  e  r  g  e  c  l  m  .  c                           \n**/
/**                                                                                \n**/
/**     Program merges N CLM files to one CLM file with N bands                    \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"
#include <sys/stat.h>

#define USAGE "Usage: mergeclm [-v] [-f] [-longheader] [-size4] in1.clm|zero [in2.clm ...] out.clm\n"

int main(int argc,char **argv)
{
  FILE **files;
  FILE *out;
  int i,iarg,year,cell,version,version_out,*nbands,setversion,numfiles,nbands_sum,index;
  Bool swapnstep;
  Byte *bvec=NULL;
  short *svec=NULL;
  int *ivec=NULL;
  float *fvec=NULL;
  double *dvec=NULL;
  void *vec=NULL;
  struct stat filestat;
  char c;
  int rc;
  Bool force,first;
  Header header,header_out;
  Bool *swap,verbose,size4;
  String id,id_out;
  /* set default values */
  force=FALSE;
  verbose=FALSE;
  size4=FALSE;
  swapnstep=FALSE;
  setversion=READ_VERSION;
  version_out=READ_VERSION;
  /* process command options */
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-f"))
        force=TRUE;
      else if(!strcmp(argv[iarg],"-v") || !strcmp(argv[iarg],"--verbose"))
        verbose=TRUE;
      else if(!strcmp(argv[iarg],"-size4"))
        size4=TRUE;
      else if(!strcmp(argv[iarg],"-longheader"))
        setversion=2;
      else
      {
        fprintf(stderr,"Error: invalid option '%s'.\n"
                USAGE,argv[iarg]);
        return EXIT_FAILURE;
      }
    }
    else
      break;

  /* calculate nbands from number of remaining arguments, last one is name of output file */
  numfiles=argc-iarg-1;
  if(numfiles<=0)
  {
    fputs("Input file missing.\n"
          USAGE,stderr);
    return EXIT_FAILURE;
  }
  files=newvec(FILE *,numfiles);
  if(files==NULL)
  {
    printallocerr("files");
    return EXIT_FAILURE;
  }
  swap=newvec(Bool,numfiles);
  if(swap==NULL)
  {
    printallocerr("swap");
    return EXIT_FAILURE;
  }
  nbands=newvec(Bool,numfiles);
  if(nbands==NULL)
  {
    printallocerr("nbands");
    return EXIT_FAILURE;
  }
  if(!force)
  {
    if(!stat(argv[argc-1],&filestat))
    {
      fprintf(stderr,"File '%s' already exists, overwrite (y/n)?\n",argv[argc-1]);
      scanf("%c",&c);
      if(c!='y')
        return EXIT_FAILURE;
    }
  }
  out=fopen(argv[argc-1],"wb");
  if(out==NULL)
  {
    fprintf(stderr,"Cannot create '%s': %s.\n",argv[argc-1],strerror(errno));
    return EXIT_FAILURE;
  }
  /* open input files and check file size for consistency */
  first=TRUE;
  nbands_sum=0;
  for(i=0;i<numfiles;i++)
  {
    if(!strcmp(argv[iarg+i],"zero"))
    {
      files[i]=NULL;
      nbands[i]=1;
      nbands_sum+=nbands[i];
    }
    else
    {
      files[i]=fopen(argv[iarg+i],"rb");
      if(files[i]==NULL)
      {
        fprintf(stderr,"Cannot open '%s': %s.\n",argv[iarg+i],strerror(errno));
        return EXIT_FAILURE;
      }
      version=setversion;
      if(freadanyheader(files[i],&header,swap+i,id,&version,TRUE))
      {
        fprintf(stderr,"Error reading header in '%s'.\n",argv[i+iarg]);
        return EXIT_FAILURE;
      }
      if(version>CLM_MAX_VERSION)
      {
        fprintf(stderr,"Error: Unsupported version %d in '%s', must be less than %d.\n",
                version,argv[i+iarg],CLM_MAX_VERSION+1);
          return EXIT_FAILURE;
      }
      if(version==4 && header.nstep>1)
      {
        if(header.nbands>1)
        {
          fprintf(stderr,"Number of bands %d and number of steps %d >1 in '%s'.\n",
                  header.nbands,header.nstep,argv[i+iarg]);
          return EXIT_FAILURE;
        }
        swapnstep=TRUE;
        nbands[i]=header.nstep;
      }
      else
        nbands[i]=header.nbands;
      nbands_sum+=nbands[i];
      if(first)
      {
        header_out=header;
        strcpy(id_out,id);
        version_out=version;
        if(size4 && version<3)
          header_out.datatype=LPJ_INT;
        first=FALSE;
      }
      else
      {
        if(header.ncell!=header_out.ncell)
        {
          fprintf(stderr,"Number of cells=%d in file '%s' differs from %d.\n",
                  header.ncell,argv[i+iarg],header_out.ncell);
          return EXIT_FAILURE;
        }
        else if(header.nyear!=header_out.nyear)
        {
          fprintf(stderr,"Number of years=%d in file '%s' differs from %d.\n",
                  header.nyear,argv[i+iarg],header_out.nyear);
          return EXIT_FAILURE;
        }
        else if(header.firstyear!=header_out.firstyear)
        {
          fprintf(stderr,"First year=%d in file '%s' differs from %d.\n",
                  header.firstyear,argv[i+iarg],header_out.firstyear);
          return EXIT_FAILURE;
        }
        else if(version>=3 && header.datatype!=header_out.datatype)
        {
          fprintf(stderr,"Data type=%s in file '%s' differs from %s.\n",
                  typenames[header.datatype],argv[i+iarg],typenames[header_out.datatype]);
          return EXIT_FAILURE;
        }
        else if(header.cellsize_lon!=header_out.cellsize_lon)
        {
          fprintf(stderr,"Cell size=%g in file '%s' differs from %g.\n",
                  header.cellsize_lon,argv[i+iarg],header_out.cellsize_lon);
          return EXIT_FAILURE;
        }
        else if(header.scalar!=header_out.scalar)
        {
          fprintf(stderr,"Scaling=%g in file '%s' differs from %g.\n",
                  header.scalar,argv[i+iarg],header_out.scalar);
          return EXIT_FAILURE;
        }
        if(verbose)
          printf("Filename: %s, bands=%d\n",argv[iarg+i],nbands[i]);
      }
      /* check file size */
      fstat(fileno(files[i]),&filestat);
      if(filestat.st_size!=header.ncell*header.nyear*nbands[i]*typesizes[header_out.datatype]+headersize(id,version))
        fprintf(stderr,"Warning: file size of '%s' does not match setting of nyear=%d and ncell=%d.\n",
                argv[iarg+i],header.nyear,header.ncell);
    }
  }
  printf("Number of nbands: %d\n",nbands_sum);
  header_out.nbands=nbands_sum;
  if(swapnstep)
  {
    version_out=4;
    header_out.nstep=header_out.nbands;
    header_out.nbands=1;
  }
  fwriteheader(out,&header_out,id_out,version_out);
  switch(header_out.datatype)
  {
    case LPJ_BYTE:
      vec=bvec=newvec(Byte,nbands_sum);
      break;
    case LPJ_SHORT:
      vec=svec=newvec(short,nbands_sum);
      break;
    case LPJ_INT:
      vec=ivec=newvec(int,nbands_sum);
      break;
    case LPJ_FLOAT:
      vec=fvec=newvec(float,nbands_sum);
      break;
    case LPJ_DOUBLE:
      vec=dvec=newvec(double,nbands_sum);
      break;
  }
  if(vec==NULL)
  {
    printallocerr("vec");
    return EXIT_FAILURE;
  }
  for(year=0;year<header_out.nyear;year++)
    for(cell=0;cell<header_out.ncell;cell++)
    {
      index=0;
      for(i=0;i<numfiles;i++)
      {
        rc=1;
        switch(header_out.datatype)
        {
          case LPJ_BYTE:
            if(files[i]==NULL)
              bvec[index]=0;
            else
              rc=fread(bvec+index,1,nbands[i],files[i]);
            break;
          case LPJ_SHORT:
            if(files[i]==NULL)
              svec[index]=0;
            else
              rc=freadshort(svec+index,nbands[i],swap[i],files[i]);
            break;
          case LPJ_INT:
            if(files[i]==NULL)
              ivec[index]=0;
            else
              rc=freadint(ivec+index,nbands[i],swap[i],files[i]);
            break;
          case LPJ_FLOAT:
            if(files[i]==NULL)
              fvec[index]=0;
            else
              rc=freadfloat(fvec+index,nbands[i],swap[i],files[i]);
            break;
          case LPJ_DOUBLE:
            if(files[i]==NULL)
              dvec[index]=0;
            else
              rc=freaddouble(dvec+index,nbands[i],swap[i],files[i]);
            break;
        }
        if(rc!=nbands[i])
        {
          fprintf(stderr,"Unexpected end of file in '%s' at year %d and cell %d.\n",argv[iarg+i],year+header_out.firstyear,cell);
          return EXIT_FAILURE;
        }
        index+=nbands[i];
      }
      /* write merged data */
      if(fwrite(vec,typesizes[header_out.datatype],nbands_sum,out)!=nbands_sum)
      {
        fprintf(stderr,"Error writing '%s': %s.\n",argv[argc-1],strerror(errno));
        return EXIT_FAILURE;
      }
    }
  free(vec);
  free(nbands);
  free(swap);
  /* close all open files */
  for(i=0;i<numfiles;i++)
    if(files[i]!=NULL)
      fclose(files[i]);
  free(files);
  fclose(out);
  return EXIT_SUCCESS;
} /* of 'main' */
